lib/connectors/memory.js

```diff
@@ -723,6 +725,62 @@ Memory.prototype.updateAttributes = function updateAttributes(model, id, data, o
   }
 };
 
+Memory.prototype.replace = function replace(model, id, data, options, cb) {
+  var self = this;
+  if (!id) {
+    var err = new Error('You must provide an id when replacing!');
+    if (cb) {
+      return cb(err);
+    } else {
+      throw err;
+    }
+  }
+
+  // Do not modify the data object passed in arguments
+  // data = Object.create(data);
+
+  this.setIdValue(model, data, id);
+  var cachedModels = this.collection(model);
+  var modelData = cachedModels && this.collection(model)[id];
+  if (modelData) {
+    var id = this.getIdValue(model, data);
+    this.collection(model)[id] = serialize(data);
+    this.saveToFile(data, function (err) {
+      cb(err, self.fromDb(model, data));
+    });
+  } else {
+    cb(new Error('Could not replace. Object with id ' + id + ' does not exist!'));
+  }
+};
+
+Memory.prototype.replaceOrCreate = function(model, data, options, callback) {
+  var self = this;
+  var idName = self.idNames(model)[0];
+  var idValue = self.getIdValue(model, data);
+  var filter = {'where': {}};
+  filter.where[idName] = idValue;
+  var nodes = self._findAllSkippingIncludes(model, filter);
+  var found = nodes[0];
+
+  if (!found) {
+    // Calling _createSync to update the collection in a sync way and 
+    // to guarantee to create it in the same turn of even loop
+    return self._createSync(model, data, function(err, id) {
+      if (err) return callback(err);
+      self.saveToFile(id, function(err, id) {
+        self.setIdValue(model, data, id);
+        callback(err, data, { isNewInstance: true });
+      });
+    });
+  } else {
+    var id = self.getIdValue(model, data);
+    self.collection(model)[id] = serialize(data);
+    self.saveToFile(data, function(err) {
+      callback(err, data, { isNewInstance: false });
```

Comment:
AFAICT, `replace` calls `self.fromDb(data)` before passing `data` to the callback. I think we should be consistent and either call it everywhere in the new code or not at all.

@raymondfeng I am not familiar with `self.fromDb`, do you happen to remember whether we should or not should apply it on the data passed to the callback?